pkg/blockbuilder/blockbuilder.go

```diff
@@ -673,6 +669,122 @@ consumerLoop:
 	return newState, nil
 }
 
+// consumePartitionSectionPullMode is for the use of scheduler based architecture.
+// startOffset is inclusive, endOffset is exclusive, and must be valid offsets and not something in the future (endOffset can be technically 1 offset in the future).
+// All the records and samples between these offsets will be consumed and put into a block.
+// The returned lastConsumedOffset is the offset of the last record consumed. It is the caller's responsibility to commit this offset.
+func (b *BlockBuilder) consumePartitionSectionPullMode(
+	ctx context.Context,
+	logger log.Logger,
+	builder *TSDBBuilder,
+	partition int32,
+	startOffset, endOffset int64,
+) (lastConsumedOffset int64, retErr error) {
+	lastConsumedOffset = startOffset
+	if startOffset >= endOffset {
+		level.Info(logger).Log("msg", "nothing to consume")
+		return
+	}
+
+	var blockMetas []tsdb.BlockMeta
+	defer func(t time.Time) {
+		// No need to log or track time of the unfinished section. Just bail out.
+		if errors.Is(retErr, context.Canceled) {
+			return
+		}
+
+		dur := time.Since(t)
+
+		if retErr != nil {
+			level.Error(logger).Log("msg", "partition consumption failed", "err", retErr, "duration", dur)
+			return
+		}
+
+		b.blockBuilderMetrics.processPartitionDuration.WithLabelValues(fmt.Sprintf("%d", partition)).Observe(dur.Seconds())
+		level.Info(logger).Log("msg", "done consuming", "duration", dur, "partition", partition,
+			"start_offset", startOffset, "end_offset", endOffset,
+			"last_consumed_offset", lastConsumedOffset, "num_blocks", len(blockMetas))
+	}(time.Now())
+
+	b.kafkaClient.AddConsumePartitions(map[string]map[int32]kgo.Offset{
+		b.cfg.Kafka.Topic: {
+			partition: kgo.NewOffset().At(startOffset),
+		},
+	})
+	defer b.kafkaClient.RemoveConsumePartitions(map[string][]int32{b.cfg.Kafka.Topic: {partition}})
+
+	level.Info(logger).Log("msg", "start consuming", "partition", partition, "start_offset", startOffset, "end_offset", endOffset)
+
+	var (
+		firstRec *kgo.Record
+		lastRec  *kgo.Record
+	)
+
+consumerLoop:
+	for recOffset := int64(-1); recOffset < endOffset-1; {
+		if err := context.Cause(ctx); err != nil {
+			return 0, err
+		}
+
+		// PollFetches can return a non-failed fetch with zero records. In such a case, with only the fetches at hands,
+		// we cannot tell if the consumer has already reached the latest end of the partition, i.e. no more records to consume,
+		// or there is more data in the backlog, and we must retry the poll. That's why the consumer loop above has to guard
+		// the iterations against the endOffset, so it retried the polling up until the expected end of the partition is reached.
```

Comment:
```suggestion
		// the iterations against the endOffset, so it retries the polling up until the expected end of the partition is reached.
```
(I know it's copy-pasted.)